rfd/0178-github-proxy.md

```diff
@@ -0,0 +1,853 @@
+message GitCommand {
+  // Metadata is a common event metadata
+  Metadata Metadata = 1 [ (gogoproto.nullable) = false, (gogoproto.embed) = true, (gogoproto.jsontag) = "" ];
+
+  // User is a common user event metadata
+  UserMetadata User = 2 [ (gogoproto.nullable) = false, (gogoproto.embed) = true, (gogoproto.jsontag) = "" ];
+
+  // ConnectionMetadata holds information about the connection
+  ConnectionMetadata Connection = 3 [ (gogoproto.nullable) = false, (gogoproto.embed) = true, (gogoproto.jsontag) = "" ];
+
+  // SessionMetadata is a common event session metadata
+  SessionMetadata Session = 4 [ (gogoproto.nullable) = false, (gogoproto.embed) = true, (gogoproto.jsontag) = "" ];
+
+  // ServerMetadata is a common server metadata
+  ServerMetadata Server = 5 [ (gogoproto.nullable) = false, (gogoproto.embed) = true, (gogoproto.jsontag) = "" ];
+
+  // CommandMetadata is a common command metadata
+  CommandMetadata Command = 6 [ (gogoproto.nullable) = false, (gogoproto.embed) = true, (gogoproto.jsontag) = "" ];
+
+  // CommandServiceType is the type of the git request like git-upload-pack or
+  // git-receive-pack.
+  string command_service_type = 8 [(gogoproto.jsontag) = "command_service_type"];
+  // Path is the Git repo path, usually <org>/<repo>.
+  string path = 9 [(gogoproto.jsontag) = "path"];
+  // Actions defines details for a Git push.
+  repeated GitCommandAction actions = 10 [(gogoproto.jsontag) = "actions,omitempty"];
+}
+
+// GitCommandAction defines details for a Git push.
+message GitCommandAction {
+  // Action type like create or update.
+  string Action = 1 [(gogoproto.jsontag) = "action,omitempty"];
+  // Reference name like ref/main/my_branch.
+  string Reference = 2 [(gogoproto.jsontag) = "reference,omitempty"];
+  // Old is the old hash.
+  string Old = 3 [(gogoproto.jsontag) = "old,omitempty"];
+  // New is the new hash.
+  string New = 4 [(gogoproto.jsontag) = "new,omitempty"];
+```
+
+#### Usage reporting
+
+There is no heartbeats for `git_server` with subkind `github` (yet).
+
+Exising `SessionStartEvent` will be expanded to include git metadata with
+`session_type` of `git`:
+```grpc
+// SessionStartGitMetadata contains additional information about git commands.         
+message SessionStartGitMetadata {                                                                  
+  // git server subkind ("github").
+  string git_type = 1;             
+  // command service type ("git-upload-pack" or "git-receive-pack").
+  string command_service_type = 2;                                                        
+}
+```
+### Security
+
+#### Provisioning GitHub username
+
+In the design described above, GitHub usernames are provided through an user
+trait. For local users, only Teleport "admins" can set the traits as modifying
+users is considered an admin action. For external users, only "admins" of the
+external identity provider, say an Okta admin, can configure the values of the
+traits.
```

Comment:
This is no longer true so probably should be removed or marked as discarded design.